### PR TITLE
Updated WooPayments MultiCurrency integration with Product Add-Ons

### LINKED ISSUE #8911: Update Product Add-Ons integration
### Background Info / Debug

SomewhereWarm adopted Product Add-ons and brought several improvements to the extension.
One of them was to **Change the way prices are presented in the cart** -  see  pdqkbG-2E8-p2

SomewhereWarm was not aware that Woo Payments integrated with Product Add-ons and didn't inform Woo Payments for these changes.

- Woo Payments Integration `get_item_data` - [See Github](https://github.com/Automattic/woocommerce-payments/blob/8aa455c724e3eeaa7a31bbe536b6266296b0592a/includes/multi-currency/Compatibility/WooCommerceProductAddOns.php#L111-L146)
- Product Add-ons `get_item_data` - [See Github](https://github.com/woocommerce/woocommerce-product-addons/blob/195901c6732aeac5e24b4011671f7aafc8ae22fb/includes/class-wc-product-addons-cart.php#L721-L802)

For this bug to be fixed, the Woo Payments integration `includes/multi-currency/Compatibility/WooCommerceProductAddOns.php` has to be updated.

See p1717513051447799-slack-C7U3Y3VMY

The issue is not only related to addons with 0 prices, as Product Add-ons has changed how all prices are presented in cart.

_This section was added by @PanosSynetos_

### Describe the bug
Product Add-Ons with no price are showing as 0.00 in the cart when WooPayments is enabled.

### To Reproduce

1. With only WooCommerce, WooPayments and Product Add-Ons active on the site, create a new product with multiple choices. One should have a flat rate of 0.00.

add-on-configuration

2. On the front-end of the site, select the "free" add-on.

product-page

3. Add this to the cart.
4. Next, repeat the steps above but with WooPayments deactivated.

### Actual behavior

When WooPayments is active, the product is added to the cart with the add-on details below the main product information. The price is displayed as 0.00 instead of not appearing at all. Repeating the process with WooPayments deactivated causes the price to be hidden as expected.

cart-zero

### Expected behavior

The price should **not** show 0.00 at the cart for "free" product add-ons.

cart-no-price

### Desktop (please complete the following information):

* OS: MacOS
* Browser: Chrome
* Version: 125.0.6422.114

### Additional context
- 8198244-zen
- p1717513051447799-slack-C7U3Y3VMY

### PR DESCRIPTION
**Pull request by jimjasson**
Fixes #8911 

#### Changes proposed in this Pull Request

This PR updates the integration between WooPayments MultiCurrency and Product Add-Ons which hasn't been reviewed after the Product Add-Ons v6.x.x releases.

One of the major changes in the newer Product Add-Ons versions is that the way add-on prices show up in the cart has changed. Add-on prices are no longer added after the add-on name, but are added after the add-on's value. Moreover, Product Add-Ons has simplified how flat fee prices show up in the cart to make it more evident to shoppers how the cart total is calculated.

Therefore, this PR: 
- updates the WooPayments MultiCurrency/Product Add-Ons integration, so MultiCurrency converts and displays add-on prices in the right location and;
- resolves an issue that made the product price in the cart change when the cart item quantity was increased. 

#### Testing instructions

First, set up a MultiCurrency environment: 
- Use WooPayments `trunk` or the latest stable version. 
- Set up WooPayments for your store. 
- Under **Payments > Settings, enable MultiCurrency** functionality. 
- Under **WooCommerce > Settings > Multi-Currency** add a couple of currencies that do not have a 1:1 conversion rate, like USD and GBP. 
- In the same screen, enable "Add a currency switcher to the Storefront theme on breadcrumb section.".

Now, let's see some bugs: 
- Install Product Add-Ons v6.9.0. 
- Create a Simple Product. 
- Go to Product Data > Add-Ons. 
- Create a 'Checkboxes' add-on. 
- Add 4 options, A, B, C, D. 
- A should have a $0 Flat Fee. 
- B should have a non-$0 Flat Fee. 
- C should have a non-$0 Quantity Based Price. 
- D should have a non-$0 Percentage Based Price. 
- View the product and select A. Add it to the cart. 
- In the cart, notice that the Flat Fee is added to the name of the add-on and is displayed as $0.00. 
- Back to the single product page, select either B or D. Add the product to the cart. 
- In the cart, change the currency. 
- Then, increase the quantity -- notice that the product price, which represents the price of the single unit, changes (this doesn't happen for products without add-ons).

Finally, let's see some fixes: 
- Grab: https://github.com/woocommerce/woocommerce-product-addons/pull/1048.
- For WooPayments, check out this branch. 
- Clear your cart and revisit the product you created. 
- Select A and add it to the cart. 
- In the cart, the $0.00 should not be visible. 
- Back to the single product page, select B. Add the product to the cart. 
- In the cart, notice that B's price shows up after the add-on value, like so (+$xx).
- Change the currency and notice that B's price is converted. 
- Increase the product quantity and ensure that the product price doesn't change. 
- Complete the order. 
- In **WooCommerce > Orders** find the order you submitted -- ensure that B's price shows up correctly there. 
- By default, quantity and percentage based prices do not show up in the cart. To make them show up to confirm that they are correct use these [snippets](https://woocommerce.com/document/product-add-ons/#display-addon-prices-cart-order).
- Back to the product page, add the product to the cart once with C and once with D selected. 
- Ensure that the prices in the cart are correct, even when changing the currency. 
- Ensure that both the product price and the add-on prices are correct when increasing the product quantity. 
- Complete the order. 
- In **WooCommerce > Orders** find the order you submitted -- ensure that C's and D's price show up correctly there. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9070` or branch name `issue-8911` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=issue-8911) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 7677ad0c4f96fd7422b927cc18c4f278d7e8b8b2
- Build time: 2024-07-17 12:18:41 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:123]: Because this defaults to `false`, the if-then condition that should handle `quantity_based` price-type is failing to resolve to `true`. Therefore, `quantity_based` price-types are being treated the same as `percentage_based` price-types.

This helps explain why the following fixes are failing:
* ❌ Increase the product quantity and ensure that the product price doesn't change.
* ❌ In **WooCommerce > Orders** find the order you submitted -- ensure that B's price shows up correctly there.
* ❌ By default, quantity and percentage based prices do not show up in the cart.
* ❌ Ensure that the prices in the cart are correct, even when changing the currency.
*  ❌ Ensure that the prices in the cart are correct for **Quantity** add-ons with **Adjust Price** set to `Quantity Based`.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Nit: Can we update this to the following?
```suggestion
			if ( class_exists( 'WC_Product_Addons_Helper' ) ) {
```

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Nit: Can we remove this `phpcs:ignore` and `@psalm-suppress` if we make the following change, above:
```php
if ( class_exists( 'WC_Product_Addons_Helper' ) ) {
```

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Nit: Same here. Can we remove these by updating the `class_exists()` check?

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Nit: Same here. Can we remove these by updating the `class_exists()` check?

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Should `WC_Product_Addons_Helper::get_product_addon_price_for_display( ... )` have two arguments? It looks like [it only has two parameters](https://github.com/woocommerce/woocommerce-product-addons/blob/195901c6732aeac5e24b4011671f7aafc8ae22fb/includes/class-wc-product-addons-helper.php#L212).

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:134]: Nit: Should we add a comment indicating that if `display` is set, it overrides value? I'm not certain it does but from my testing, it looks like it does.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Nit: Would it be more accurate to update this function's name to `maybe_convert_flat_fees`? If so, we can also update the comment.

For example, the flat fee won't be converted if the default currency is already being used.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:273]: Just wanted to confirm: Should the timestamp overwrite what was added above (e.g. `$addon['value']`) or should it be appended?

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:303]: We may want to consider handling things more gracefully when `$price` is not defined. Which can happen if `class_exists( '\WC_Product_Addons_Helper' )` resolves to `false`.

**Review by lovo-h — CHANGES_REQUESTED** [review]
Thank you for addressing these issues 🙏 

The code changes are looking good, but I was not able to observe some of the behavior, as described. I also left a few nit comments that we may want to apply before merging 😎

### Testing

**Bugs**
* ✅ In the cart, notice that the Flat Fee is added to the name of the add-on and is displayed as $0.00.
* ❌ Then, increase the quantity -- notice that the product price, which represents the price of the single unit, changes (this doesn't happen for products without add-ons).
	* I did notice this bug, but only with the changes in this PR.

**Fixes**
* ✅ In the cart, the $0.00 should not be visible.
* ✅ In the cart, notice that B's price shows up after the add-on value, like so (+$xx).
* ✅ Change the currency and notice that B's price is converted.
* ❌ Increase the product quantity and ensure that the product price doesn't change.
	* The product price, which represents the price of the single unit, changes.
* ❌ In **WooCommerce > Orders** find the order you submitted -- ensure that B's price shows up correctly there.
	* ❌ The product price is the total divided by the quantity.
	* The add-on doesn't display the price next to it. But I don't know if it should.
* ❌ By default, quantity and percentage based prices do not show up in the cart.
	* They show up even without the snippets activated.
* ❌ Ensure that the prices in the cart are correct, even when changing the currency.
	* ❌ Add-on C - The base price plus the add-on price adds up to the total price, but the displayed add-on price is incorrect.
	* ✅ Add-on D - The base price plus the add-on price add up to the total price and the displayed add-on price is correct.
* ✅ Ensure that both the product price and the add-on prices are correct when increasing the product quantity.

**Other Tests**
* ❌ Ensure that the prices in the cart are correct for Quantity add-ons with **Adjust Price** set to `Quantity Based`.

**Comment by jimjasson** [comment]
Thank you @lovo-h for the thorough review! 🙏

I pushed some fixes both on this branch and on: https://github.com/woocommerce/woocommerce-product-addons/pull/1048, so feel free to pull both branches and have another look! 

> Then, increase the quantity -- notice that the product price, which represents the price of the single unit, changes (this doesn't happen for products without add-ons).
I did notice this bug, but only with the changes in this PR.

This bug is a bit random, but you are right that it was not completely fixed by the time you tested this PR.  

With the latest stable WooPayments and Product Add-Ons versions active, I am able to replicate this issue by:
- creating a product that costs $20 and has a $10 flat fee add-on, 
- adding the product with the add-on selected to the cart, 
- changing the currency to GBP and; 
- increasing the product quantity. 

https://github.com/user-attachments/assets/887600ce-e66f-410c-af60-5760fba93819

Don't worry too much if you cannot replicate it -- let's make sure that you don't see it after the latest changes.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:123]: Correct! I have updated the if-checks so that they are specific for percentage-based add-ons.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Good idea! Resolved.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: Correct, resolved!

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php]: This part of code is no longer needed after the latest commits.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:273]: Yes, this is 1-1 identical with what Product Add-Ons does. Since the WooPayments function overrides everything coming from Product Add-Ons, we should initialize the add-on value to the timestamp for Date Picker add-ons.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:303]: I thought about this, but couldn't find a case where the `WC_Product_Addons_Helper` wouldn't be defined -- do you have a specific case in mind? 

Right now, if this class is not defined, the the add-on value will be the same as the one set by Product Add-Ons, i.e. not converted by the selected currency. Would you prefer to have a different fallback here?

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:134]: This is also 1-1 identical to how Product Add-Ons [handles](https://github.com/woocommerce/woocommerce-product-addons/blob/trunk/includes/class-wc-product-addons-cart.php#L784) the `display` key -- in general, the `display` key is not actively used, but is there for backwards compatibility.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (lovo-h) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:303]: Although it's unlikely, this is one scenario where this may happen:
* We remove `WC_Product_Addons_Helper` in a future version of `woocommerce-product-addons`.
* The merchant installs the future version of `woocommerce-product-addons`.
* The merchant installs this version of WooPayments.

Although this is possible, it's unlikely, so we can proceed without making modifications.

**Review by lovo-h — APPROVED** [review]
Great work! Code changes look good and test well 👍 

**Fixes**
* 🟢 In the cart, the $0.00 should not be visible.
* 🟢 In the cart, notice that B's price shows up after the add-on value, like so (+$xx).
* 🟢 Change the currency and notice that B's price is converted.
* ✅ Increase the product quantity and ensure that the product price doesn't change.
* ✅ In **WooCommerce > Orders** find the order you submitted -- ensure that B's price shows up correctly there.
	* ❓ The add-on doesn't display the price next to it. I'm not sure if it should.
* ✅ By default, quantity and percentage based prices do not show up in the cart. To make them show up to confirm that they are correct use these [snippets](https://woocommerce.com/document/product-add-ons/#display-addon-prices-cart-order).
* ✅ Ensure that the prices in the cart are correct, even when changing the currency.
* 🟢 Ensure that both the product price and the add-on prices are correct when increasing the product quantity.

**Other Tests**
* ✅ Ensure that the prices in the cart are correct for Quantity add-ons with **Adjust Price** set to `Quantity Based`.
* ✅ `npm run test:php`.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:125]: This might be good enough here, but comparing floats for equality might cause subtle problems. Perhaps 0 is a special case here, but I think price can often be filtered and calculated, so maybe not totally safe?

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:127]: Same here with comparing floats for equality.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:162]: Out of scope for this PR, but interesting that we update the product price with percentage-based add-ons, but don't update it for quantity-based add-ons.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [inline includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:111]: Since this method is very similar to \WC_Product_Addons_Cart::get_item_data, would it be worth it to add a comment to PAO to reflect changes in \WC_Product_Addons_Cart::get_item_data here in Woo Payments, too, so we can avoid things being out of sync in the future? (and similarly with other methods here that have counterparts in PAO)

Alternatively, we could refactor \WC_Product_Addons_Cart::get_item_data (or extract the calculation part out to a separate method) and call it from WooPayments, it's essentially a static method anyway, right?

**Review by peterfabian — COMMENTED** [review]
Hi @jimjasson 

I added a couple of comments, but I'm trying to test it on the jurassic ninja site and having problems getting the numbers to add up.

Using a simple product with price 21 and the following addons:
Screenshot 2024-07-17 at 13 45 25

The calculation in the cart differs from what I see on the single product page even with the default currency ($):
Screenshot 2024-07-17 at 13 48 02
Screenshot 2024-07-17 at 13 48 22

It seems to me the total should be $57.30, with the product price being $47.30.

The calculation on the single product page is what I'd expect and it also aligns with what I see in PAO trunk locally:
Screenshot 2024-07-17 at 13 48 58

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:125]: In the Product Add-Ons repo, we do this instead: https://github.com/woocommerce/woocommerce-product-addons/blob/trunk/includes/class-wc-product-addons-cart.php#L769, but Woo Payments repo doesn't allow non-strict comparisons and the `float` method was previously used here: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Compatibility/WooCommerceProductAddOns.php#L237

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:162]: Yes, this happens because the percentage amount needs to be calculated on top of the converted product price. Flat fees and quantity based add-ons are added on top of the product price and they don't need the product price to make any calculation about the actual add-on value.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (jimjasson) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:111]: Added some comments [here](https://github.com/woocommerce/woocommerce-product-addons/pull/1048/commits/995c188c43c8a83e67c92c6a2e4a51785db38abb).

**Review by peterfabian — COMMENTED** [review]
After the changes, the numbers now match.  🎉 

I wasn't able to check out with product qty > 1 due to the rounding problem with Level 3 data, but I will create a separate issue for that, since it seems that this problem also happens with the latest stable WooPayments.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:162]: Thanks for the explanation.

File: includes/multi-currency/Compatibility/WooCommerceProductAddOns.php
Comment (peterfabian) [reply includes/multi-currency/Compatibility/WooCommerceProductAddOns.php:111]: Thanks!

**Review by peterfabian — APPROVED** [review]
Tested and reviewed from different angles, seems to be working as expected now. 

The remaining issue was created as https://github.com/Automattic/woocommerce-payments/issues/9114.

**Comment by jimjasson** [comment]
Thank you Peter! Will merge this PR after: https://github.com/woocommerce/woocommerce-product-addons/pull/1048 gets released, by the end of July.

**Pull request merged**
